test/Test/Data/Array/NonEmpty.purs

```diff
@@ -397,6 +397,22 @@ testNonEmptyArray = do
   log "traverse1 should work"
   assert $ traverse1 Just (fromArray [1, 2, 3, 4]) == NEA.fromArray [1, 2, 3, 4]
 
+  log "transpose swaps rows and columns for a regular two-dimension array"
+  assert $ NEA.transpose (fromArray [ (fromArray [1,2,3]), (fromArray [4,5,6]), (fromArray [7,8,9])]) ==
+                         (fromArray [ (fromArray [1,4,7]), (fromArray [2,5,8]), (fromArray [3,6,9])])
+  
+  log "transpose skips elements when rows don't match"
+  assert $ NEA.transpose (fromArray [ (fromArray [10,11]), (fromArray [20]), (fromArray [30,31,32])]) ==
+                         (fromArray [ (fromArray [10,20,30]), (fromArray [11,31]), (fromArray [32])])
+
+  log "transpose' swaps rows and columns for a regular two-dimension array"
+  assert $ NEA.transpose' (fromArray [[1,2,3], [4,5,6], [7,8,9]]) ==
+                          (fromArray [[1,4,7], [2,5,8], [3,6,9]])
+  
+  log "transpose' skips elements when rows don't match"
+  assert $ NEA.transpose' (fromArray [[10,11], [20], [30,31,32]]) ==
+                          (fromArray [[10,20,30], [11,31], [32]])
+
```

Comment:
Can another test be added here for `NEA.transpose' (fromArray [[]]) == (fromArray [[]])`?

Reply:
Hmm, this test fails - it produces ```(fromArray [])```.  Perhaps we need to special-case it?

Reply:
Oh, I think my example is just wrong. `Array` has the same behavior in its test: https://github.com/purescript/purescript-arrays/blob/master/test/Test/Data/Array.purs#L280

Reply:
I guess the test should be `NEA.transpose' (fromArray [[]]) == (fromArray [])`

Reply:
Surely ```fromArray []``` is an empty ```NonEmptyArray``` and thus eminently unsafe.

Reply:
Oh gosh... :man_facepalming: Thanks!

Then the implementation should be
```purs
transpose' :: forall a. NonEmptyArray (Array a) -> Maybe (NonEmptyArray (Array a))
transpose' = fromArray <<< A.transpose <<< coerce
```

Reply:
I'm not entirely happy with this.  From my perspective, ```transpose' (fromArray [[]])``` is a very odd thing to do anyway and I wouldn't have thought it would be entirely incorrect for it to be an idempotent operation.  But the cost of doing what you suggest  is to make all the normal cases more awkward by wrapping the result in the ```Maybe```.

Reply:
Awkward or not, it is correct. And if that's the definition, then it's the same as:
```purs
fromArray <<< A.transpose <<< NEA.toArray
```
So, I would argue now that we shouldn't implement that here.

Reply:
I'll happily do what you suggest if it is the correct definition.  Unless I hear otherwise from @garyb today I'll do it this way - I don't feel strongly about it in any case.

(However just to expand my approach a bit, I'd argue that the minimal result for ```transpose Array (Array a)``` is ```[]```, that for ```NonEmptyArray (NonEmptyArray a)``` is ```[[something]]``` and that for ```NonEmptyArray (Array a)``` is ```[[]]```.  There seems to be no equivalent function on ```NonEmptyList```` in the Haskell library. which might indicate correctness.)

Reply:
The result of transposing `[[]]` being `[[]]` does sound pretty reasonable to me (in both non-or-empty cases), but I guess if we have evidence from elsewhere that that's not normal, then yeah having `transpose'` fail with `Nothing` is probably the best path forward.